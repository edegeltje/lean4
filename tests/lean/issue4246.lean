/-!
  # Allow induction and cases to find instance target arguments which are not found by unification

  When unification with explicit targets fails to supply all implicit target arguments,
  as a last effort to provide instance-implicit arguments, the elaborator will try synthesis.
-/
class Foo (α : Type) where
  x : Nat

instance : Foo Nat where
  x := 0

/-!
  In this example, we have an induction principle Foo.induction where `(α : Type)` and `[ohno : Foo α]` are targets.
  The `ohno` target cannot be solved by unification of a later argument.
  However, we *are* able to solve for the argument by instance synthesis, so we should try to do so.
  When making this attempt fails, an error should be thrown which makes clear that we need this instance.
-/
section

@[elab_as_elim]
def Foo.induction (P : ∀ (α : Type) [Foo α], Prop) (nat : P Nat) (α : Type) [ohno : Foo α] : P α := sorry

/-! elaboration is able to infer [Foo α] -/
example (α : Type) [Foo α] : α = Nat := by
  induction α using Foo.induction with
  | nat => rfl


/-! elaboration synthesizes the right instance -/
set_option pp.explicit true in
example (α : Type) {noninst : Foo α} [inst:Foo α] : noninst.x = inst.x := by?
  induction α using Foo.induction with
  | nat => rfl

/-! elaboration shows a "failed to synthesize" error when the instance can't be synthesized -/
example (α : Type) : α = Nat := by
  induction α using Foo.induction with
  | nat => rfl
end

structure Bar (α : Type) [Foo α] : Type where
  hy : Foo.x α = 2

@[elab_as_elim]
def Bar.induction (P : ∀ (α : Type) [Foo α] (b: Bar α), Prop)
  (α : Type) [ohno : Foo α] (x : Bar α) (nat : ∀ (x : Bar Nat), P Nat x) : P α x := sorry


set_option pp.all true in
/-! unification provides `noninst` as correct target instead of finding `inst` via synthesis -/
example (α : Type) {noninst : Foo α} [inst : Foo α] (x:@Bar α noninst) : Nonempty (@Bar α noninst) := by
  induction α, x using Bar.induction with
  | nat x => exact Nonempty.intro x
