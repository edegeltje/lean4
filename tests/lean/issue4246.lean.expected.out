issue4246.lean:22:4-22:17: warning: declaration uses 'sorry'
issue4246.lean:31:11-31:14: error: tactic 'rfl' failed, the left-hand side
  @Foo.x Nat noninst
is not definitionally equal to the right-hand side
  @Foo.x Nat instFooNat
case nat
α : Type
noninst : Foo Nat
⊢ @Eq Nat (@Foo.x Nat noninst) (@Foo.x Nat instFooNat)
Try this: @Foo.induction (fun α [inst : Foo α] => ∀ {noninst : Foo α}, @Eq Nat (@Foo.x α noninst) (@Foo.x α inst))
  (fun {noninst} =>
    sorryAx (∀ (tag : Lean.Name), @Eq Nat (@Foo.x Nat noninst) (@Foo.x Nat instFooNat)) true
      `issue4246.31.8.31.14.8.14._sorry._@.issue4246._hyg.104)
  α inst noninst
issue4246.lean:39:2-40:14: error: failed to synthesize
  Foo α

Additional diagnostic information may be available using the `set_option diagnostics true` command.
issue4246.lean:47:4-47:17: warning: declaration uses 'sorry'
